### PR TITLE
Fix unordered_flat_map equals, add missing methods

### PR DESCRIPTION
**Pull request by Neverlord**
I've made a pass over our `unordered_flat_map`.

### CODE REVIEW

File: libcaf_core/caf/unordered_flat_map.hpp [hunk 1]
```diff
@@ -183,6 +178,21 @@ class unordered_flat_map {
...
+                                             mapped_type val) {
+    if (auto i = find(key); i != end()) {
+      i->second = std::move(val);
+      return {i, false};
+    }
+    xs_.emplace_back(value_type{key, std::move(val)});
+    return {xs_.end() - 1, true};
+  }
+
+  iterator insert_or_assign(const_iterator, const key_type& key,
```
Comment (riemass) [inline libcaf_core/caf/unordered_flat_map.hpp:191]: Same as above, hint would now be deprecated?

**Review by riemass — CHANGES_REQUESTED** [review]
I'm all in for removing `operator<`, doesn't really make sense in an unordered container. I've made some minor suggestions below.

File: libcaf_core/caf/unordered_flat_map.hpp [hunk 2]
```diff
@@ -158,13 +158,8 @@ class unordered_flat_map {
...
   }
 
-  iterator insert(iterator hint, value_type x) {
-    return insert(static_cast<const_iterator>(hint), std::move(x));
-  }
-
-  iterator insert(const_iterator hint, value_type x) {
-    auto i = find(x.first);
-    return i == end() ? xs_.insert(hint, std::move(x)) : i;
+  iterator insert(const_iterator, value_type x) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/unordered_flat_map.hpp:161]: The class is supposed to be a drop-in replacement for STL map types. Even though we don't do something with the hint, it's still imperative that we provide the STL interface. 🙂

File: libcaf_core/caf/unordered_flat_map.test.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  SECTION("calling erase with a non-existing key does nothing") {
+    check_eq(xs.size(), 4u);
+    check_eq(xs.erase(5), 0u);
+    check_eq(xs.size(), 4u);
+  }
+}
+
+TEST("element lookup") {
+  int_map xs{{1, 10}, {2, 20}, {3, 30}, {4, 40}};
+  SECTION("at() accesses existing elements or inserts new ones") {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/unordered_flat_map.test.cpp:141]: Yeah, that comment is blatantly wrong. Calling `at()` will throw if no such element exists. Good catch.

File: libcaf_core/caf/unordered_flat_map.hpp [hunk 4]
```diff
@@ -195,36 +205,35 @@ class unordered_flat_map {
 
   size_type erase(const key_type& x) {
     auto pred = [&](const value_type& y) { return x == y.first; };
-    auto i = std::remove_if(begin(), end(), pred);
-    if (i == end())
-      return 0;
-    erase(i);
-    return 1;
+    if (auto i = std::remove_if(begin(), end(), pred); i != end()) {
+      erase(i);
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/unordered_flat_map.hpp]: Done.

**Review by riemass — APPROVED** [review]

**Pull request merged**